### PR TITLE
LEAF-2034 Add File Tracking to Editors

### PR DESCRIPTION
**Pull request by jrodcode**
This feature enables users to make changes to different sections of the admin panel and track multiple versions of edited files. The system allows for seamless merging of edited files in case of content changes. This streamlines the editing process, prevents errors, and ensures that all changes are recorded and stored for future reference.

### CODE REVIEW

**Review by jampaul3 — CHANGES_REQUESTED** [review]

File: libs/logger/DataActionLogger.php [hunk 1]
```diff
@@ -71,20 +74,52 @@ public function getHistory($filterById, $filterByColumnName, $logType){
...
+    {
         $sql = "SELECT `value`
                 FROM `data_log_items`
                 WHERE `column` = 'groupID'
                 GROUP BY `value`;";
         return  $this->db->query_kv($sql, 'value', 'value', array());
     }
 
-    function fetchLogData($filterById, $filterByColumnName, $logType){
+    function fetchLogData($filterById, $filterByColumnName, $logType)
```
[the remark this answers is not in the document]
Comment (jrodcode) [reply libs/logger/DataActionLogger.php]: Fixed, thank you for letting me know!

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by jampaul3 — APPROVED** [review]

File: LEAF_Request_Portal/api/index.php [hunk 2]
```diff
@@ -176,4 +176,55 @@
     echo $TemplateReportsController->handler($action);
 });
 
+
+$controllerMap->register('emailTemplateFileHistory', function () use ($p_db, $login, $action) {
```
Comment (pete-nerantzinis) [inline LEAF_Request_Portal/api/index.php:180]: For EmailTemplateController and TemplateFileHIstoryController there should only be 1 controller for each, these are all considered duplicates.

File: docker/mysql/db/db_upgrade/portal/Update_RMC_DB_2023030900-2023040700.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+   `file_parent_name` text,
+   `file_name` text,
+   `file_path` text,
+   `file_size` mediumint(9) DEFAULT NULL,
+   `file_modify_by` text,
+   `file_created` text,
+   PRIMARY KEY (`file_id`)
+ ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+ UPDATE `settings` SET `data` = '2023040600' WHERE `settings`.`setting` = 'dbversion';
```
Comment (pete-nerantzinis) [inline docker/mysql/db/db_upgrade/portal/Update_RMC_DB_2023030900-2023040700.sql:14]: your db file is labeled 2023040700 make sure this matches

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 4]
```diff
@@ -50,247 +360,545 @@
...
+        loadContent();
+    }
+
+
+    var currentFile = '';
+    var currentFileContent = '';
+
+    function loadContent(file) {
+        if (file == undefined) {
+            file = currentFile;
```
Comment (raphaelfretter) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:792]: May be worthwhile to alert the user that the file could not be loaded.

**Review by raphaelfretter — CHANGES_REQUESTED** [review]

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl (hunk 4 above)
Comment (jrodcode) [reply LEAF_Request_Portal/admin/templates/mod_templates.tpl:792]: Will add that, thank you!

File: docker/mysql/db/db_upgrade/portal/Update_RMC_DB_2023030900-2023040700.sql (hunk 3 above)
Comment (jrodcode) [reply docker/mysql/db/db_upgrade/portal/Update_RMC_DB_2023030900-2023040700.sql]: Have made the changes.

File: LEAF_Request_Portal/api/index.php (hunk 2 above)
Comment (jrodcode) [reply LEAF_Request_Portal/api/index.php:179]: Have made the changes

**Review by shaneodd — APPROVED** [review]

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 5]
```diff
@@ -50,247 +360,550 @@
...
+        } else {
+            data = codeEditor.getValue();
+        }
+        $.ajax({
+                type: 'POST',
+                data: {CSRFToken: '<!--{$CSRFToken}-->',
+                file: data
+            },
+            url: '../api/templateEditor/_' + currentFile,
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:441]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 6]
```diff
@@ -50,247 +360,550 @@
...
+        } else {
+            data = codeEditor.getValue();
+        }
+        $.ajax({
+                type: 'POST',
+                data: {CSRFToken: '<!--{$CSRFToken}-->',
+                file: data
+            },
+            url: '../api/templateFileHistory/_' + currentFile,
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:473]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 7]
```diff
@@ -50,247 +360,550 @@
...
-			    $('#btn_compare').css('display', 'none');
-			}
-
-            var time = new Date().toLocaleTimeString();
-            $('#saveStatus').html('<br /> Last saved: ' + time);
-            currentFileContent = data;
-            if(res != null) {
-                alert(res);
+    function saveFileHistory() {
+        var data = '';
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:461]: Is "data" internal (so should be let) or global?

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 8]
```diff
@@ -50,247 +360,550 @@
...
+    function restore() {
+        dialog.setTitle('Are you sure?');
+        dialog.setContent('This will restore the template to the original version.');
+
+        dialog.setSaveHandler(function() {
+            $.ajax({
+                type: 'DELETE',
+                url: '../api/templateEditor/_' + currentFile + '?' +
+                    $.param({'CSRFToken': '<!--{$CSRFToken}-->'}),
+                    success: function() {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:488]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 9]
```diff
@@ -50,247 +360,550 @@
...
+    function compare() {
+        $('.CodeMirror').remove();
+        $('#codeCompare').empty();
+        $('#btn_compare').css('display', 'none');
+        $('#btn_compareStop').css('display', 'block');
+
+        $.ajax({
+            type: 'GET',
+            url: '../api/templateEditor/_' + currentFile + '/standard',
+            success: function(standard) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:509]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 10]
```diff
@@ -50,247 +360,550 @@
...
+        });
+        $.ajax({
+            type: 'GET',
+            url: '../api/templateCompareFileHistory/_' + fileName,
+            dataType: 'json',
+            cache: false,
+            success: function(res) {
+                var filePath = '';
+                var fileParentFile = '';
+                for (var i = 0; i < res.length; i++) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:650]: "(let i = 0..."

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 11]
```diff
@@ -50,247 +360,550 @@
...
+            $('.CodeMirror-linebackground').css({
+                'background-color': '#8ce79b !important'
+            });
+        });
+        $.ajax({
+            type: 'GET',
+            url: '../api/templateCompareFileHistory/_' + fileName,
+            dataType: 'json',
+            cache: false,
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:647]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates.tpl [hunk 12]
```diff
@@ -50,247 +360,550 @@
...
+
+        initEditor();
+        currentFile = file;
+        $('#codeContainer').css('display', 'block');
+        $('#controls').css('visibility', 'visible');
+        $('#filename').html(file.replace('.tpl', ''));
+        $.ajax({
+            type: 'GET',
+            url: '../api/templateEditor/_' + file,
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates.tpl:809]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates_email.tpl [hunk 13]
```diff
@@ -156,366 +499,651 @@
...
+                file: data,
+                subjectFile: subject,
+                subjectFileName: currentSubjectFile,
+                emailToFile: emailToData,
+                emailToFileName: currentEmailToFile,
+                emailCcFile: emailCcData,
+                emailCcFileName: currentEmailCcFile
+            },
+            url: '../api/emailTemplateFileHistory/_' + currentFile,
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates_email.tpl:593]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates_email.tpl [hunk 14]
```diff
@@ -156,366 +499,651 @@
...
+        $('.CodeMirror').remove();
+        $('#codeCompare').empty();
+        $('#subjectCompare').empty();
+        $('#btn_compare').css('display', 'none');
+        $('#btn_compareStop').css('display', 'block');
+        // Get default email template fields
+        $.ajax({
+            type: 'GET',
+            url: '../api/emailTemplates/_' + currentFile + '/standard',
+            success: function(standard) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates_email.tpl:654]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates_email.tpl [hunk 15]
```diff
@@ -156,366 +499,651 @@
...
+            $('.CodeMirror-linebackground').css({
+                'background-color': '#8ce79b !important'
+            });
+        });
+        $.ajax({
+            type: 'GET',
+            url: '../api/templateCompareFileHistory/_' + fileName,
+            dataType: 'json',
+            cache: false,
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates_email.tpl:823]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates_email.tpl [hunk 16]
```diff
@@ -156,366 +499,651 @@
...
+            $('#divSubject, #divEmailTo, #divEmailCc').hide().attr('disabled', 'disabled');
+            subjectEditor.setOption("readOnly", true);
+        } else {
+            $('#subject, #emailLists, #emailTo, #emailCc').show();
+            $('#divSubject, #divEmailTo, #divEmailCc').show().removeAttr('disabled');
+        }
+        $.ajax({
+            type: 'GET',
+            url: '../api/emailTemplates/_' + file,
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates_email.tpl:1005]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates_email.tpl [hunk 17]
```diff
@@ -156,366 +499,651 @@
...
+    function viewHistory() {
+        dialog_message.setContent('');
+        dialog_message.setTitle('Access Template History');
+        dialog_message.show();
+        dialog_message.indicateBusy();
+        $.ajax({
+            type: 'GET',
+            url: 'ajaxIndex.php?a=gethistory&type=emailTemplate&id=' + currentFile,
+            dataType: 'text',
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates_email.tpl:1092]: fail/catch

File: LEAF_Request_Portal/admin/templates/mod_templates_email.tpl [hunk 18]
```diff
@@ -156,366 +499,651 @@
...
+            'confirm_button_save', 'confirm_button_cancelchange');
+        initEditor();
+        $(window).on('resize', function() {
+            updateEditorSize();
+        });
+        // Get initial email tempates for page from database
+        $.ajax({
+            type: 'GET',
+            url: '../api/emailTemplates',
+            success: function(res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/admin/templates/mod_templates_email.tpl:1120]: fail/catch

**Review by Pelentan — APPROVED** [review]
I changed my mind.  If this works, we'll just push it out.  We'll catch the tech debt later.

**Review by maxrichardjr — APPROVED** [review]

**Review by Pelentan — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by AmandaSumner — APPROVED** [review]

**Review by jampaul3 — APPROVED** [review]

**Review by pete-nerantzinis — CHANGES_REQUESTED** [review]
Get rid of those .gitkeep files please or add them to the ignore

**Comment by jrodcode** [comment]
>

**Review by aerinkayne — CHANGES_REQUESTED** [review]
In the template editor, I noticed that the save button is not present when comparing to original.  We will want to retain the ability to pull in specific changes from there.

File: LEAF_Request_Portal/admin/templates/mod_templates_reports.tpl [hunk 19]
```diff
@@ -3,265 +3,970 @@
...
+            url: '../api/reportTemplates/fileHistory/_' + currentFile,
+            success: function(res) {
+                console.log("File history has been saved.");
+                getFileHistory(currentFile);
+            }
+        });
+    }
+
+    function formatFileSize(bytes, threshold = 1024) {
+        if (bytes < threshold) {
```
Comment (raphaelfretter) [inline LEAF_Request_Portal/admin/templates/mod_templates_reports.tpl:495]: This could probably be done with a while loop but it's fine for now.

**Review by raphaelfretter — COMMENTED** [review]
left comments

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

File: LEAF_Request_Portal/admin/templates/mod_templates_reports.tpl (hunk 19 above)
Comment (jrodcode) [reply LEAF_Request_Portal/admin/templates/mod_templates_reports.tpl:495]: Have made the change to a while loop

**Review by pete-nerantzinis — APPROVED** [review]

**Pull request merged**